Add `DockerWaitHealthyContainer` task

gabrielhuff opened the pull request:
## Work Summary

- Add `DockerWaitHealthyContainer` task
  - Create functional tests
    - Create test class
    - Create successful execution test
    - Create timeout test
    - Create quick stop test
    - Create not running test
    - Add cleanup logic (revertable)
  - Create task
    - Create task class
    - Define I/O
    - Create utility
    - Implement task logic
    - Move to `extras` subpackage (revertable)
- Correction 1
  - Update interface: expose polling behavior as API
    - Add reactive tests
      - Create periodic emission test
      - Create last successful emission test 
    - Expose polling cycle as input
    - Invoke `onNext` at each cycle passing the health state string
  - Update implementation
    - Check based on command instead of client
    - Document IO
    - Rewrite polling `while` loop
- Correction 2
  - Update README file

## Notes

- I'm destroying the images and containers created by the tests so that the machine running them doesn't get flooded. I'm not sure if there is a reason why this is not being done on the other tests (maybe to isolate the features being tested or make the code more readable), but this behavior can be removed by reverting commit deb37eb.
- The new task was created under the `com.bmuschko.gradle.docker.tasks.container.extras` package as suggested by @cdancy. It can be moved to the container package by undoing commit c97f9ac.
- I changed the way timeout works (comparing to the snippet from the issue) so that it looks similar to the `DockerWaitContainer` task. The following changes were made:
  -  The `timeout` input can now be `null` (no timeout), which now is the default value
  -  The `checkInterval` input was removed in order to make the API more simple.

## Open Tasks

- [x] **Update README** - Since the task belongs to a new package, I'm not sure about how to update the README file.

orzeh commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/extras/DockerWaitHealthyContainer.groovy:
```diff
@@ -0,0 +1,51 @@
+
+    @Input
+    @Optional
+    Integer timeout
```
Please add doc about time unit

gabrielhuff replied:
Ops! Completely forgot about this one

orzeh commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/extras/DockerWaitHealthyContainer.groovy:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private boolean check(dockerClient) {
+        def command = dockerClient.inspectContainerCmd(getContainerId())
```
Maybe we should reuse command instance instead of recreating it every 100 milliseconds?

gabrielhuff replied:
Good idea. Fixed on 34f10e2

orzeh commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/extras/DockerWaitHealthyContainer.groovy:
```diff
@@ -0,0 +1,51 @@
+        def timeout = { deadline ? System.currentTimeMillis() > deadline : false }
+        while(!timeout()) {
+            if (check(dockerClient)) return
+            sleep(100)
```
I would allow user to define it (via `checkInterval`) and set the default value for at least 500 milliseconds.

gabrielhuff replied:
The reason why I removed `checkInterval` was because it felt like it was an implementation detail. Currently we are polling the container health status, but maybe the docker API allows us to be notified via callbacks on the future. Anyway, fixed on 161ec4b

cdancy commented:
Let's take into account the use of `onNext` callback and allow the user to get ..
 Something... With each iteration of this wait look. Perhaps pass in the state of the container?

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/extras/DockerWaitHealthyContainer.groovy:
```diff
@@ -0,0 +1,51 @@
+            if (check(dockerClient)) return
+            sleep(100)
+        }
+        if (!check(dockerClient)) throw new GradleException("Health check timeout expired")
```
Feels like we can optimize this somehow. We shouldn't have to make another follow up call once outside of the loop.

gabrielhuff replied:
I rewrote this on 7fbfea8

cdancy commented on src/main/groovy/com/bmuschko/gradle/docker/tasks/container/extras/DockerWaitHealthyContainer.groovy:
```diff
@@ -0,0 +1,51 @@
+        if (!check(dockerClient)) throw new GradleException("Health check timeout expired")
+    }
+
+    private boolean check(dockerClient) {
```
Lets rename this to `isContainerHealthy`.

gabrielhuff replied:
Since the logic changed on 7fbfea8, I'm sticking with `check` since it's calling `onNext` and verifying the timeout all at once

cdancy commented:
I second @orzeh thoughts: very good start. Thanks for the detailed write up, clean code, and well written functional tests. I added some additional comments as well to address. Thanks!

gabrielhuff commented:
Thanks again for the feedback. I updated PR description with the new changes (under the 'Correction 1' task).

I followed @cdancy 's interesting suggestion about the `onNext` callback. Also fixed some implementation related stuff.

cdancy commented:
LGTM as well. @gabrielhuff as far as editing our README I would create another section entitled `Extras`, just below `Networks`, and add a single rowed table with a link to this task and a brief description. Once that is done than we can merge.

cdancy commented:
LGTM. As soon as this PR goes through I will merge.

gabrielhuff commented:
@cdancy @orzeh Thanks for the opportunity guys. Keep up the good work, this plugin is amazing!
